### ISSUE TITLE
bin/magento setup:upgrade causes metadata locks

### ISSUE THREAD

**Issue by Flowlance** [magento/magento2#19220]
### Preconditions (*)
1. Magento 2.2.6
2. MySQL 5.7.23 (Amazon RDS)
3. Jenkins
4. A huge store with 250000 products, and each product has more than 200 attributes.

### Steps to reproduce (*)
1. We're using Jenkins for continuous integration. 
2. When running "bin/magento setup:upgrade" from the jenkins build server, it takes a really long time (sometimes 1 hour+).
3. After viewing MySql server processes, I can see that there's a lot queries with "Waiting for table metadata lock".
4. It seems to be stuck on "Magento_Indexer":

```
Data post-updates:
Module 'Magento_Store':
Module 'Amasty_Base':
Module 'Amasty_Geoip':
Module 'Amasty_Scroll':
Module 'Amasty_ShopbyBase':
Module 'Magento_Directory':
Module 'Amasty_ShopbyBrand':
Module 'Magento_Theme':
Running data recurring...
Module 'Magento_Eav':
Module 'Magento_Customer':
Module 'Magento_Indexer':
```

### Expected result (*)
1. "bin/magento setup:upgrade" shouldn't take so long.
2. No waiting for metadata locks.

### Actual result (*)

1. 
![screenshot 2018-11-15 at 10 36 29](https://user-images.githubusercontent.com/3727974/48544000-ff572200-e8c2-11e8-9a48-3f8d74ca079e.png)
2. 
![screenshot 2018-11-15 at 10 43 23](https://user-images.githubusercontent.com/3727974/48544322-9e7c1980-e8c3-11e8-8de6-2f0aa4fcb663.png)
3. We're getting metadata locks and it can takes over an hour for them to disappear.

**Comment by magento-engcom-team** [comment]
Hi @Flowlance. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@Flowlance do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @TomashKhamlai. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by TomashKhamlai** [comment]
@Flowlance do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?

Could you please share your experience? How the products and attributes were moved to vanilla Magento Open Source instance?

**Comment by Flowlance** [comment]
@TomashKhamlai Hi. I cannot reproduce this on a vanilla installation. setup:upgrade is very quick there.
What's the best way to copy this number of products over to the vanilla instance? 
The system->export doesn't seem to handle this amount very well. It crashes.

**Comment by orlangur** [comment]
@Flowlance you can try using vanilla instance with the copy of this database.

**Comment by advcha** [comment]
@Flowlance I have an experience on this problem. I don't know exactly why this happened. Probably you have huge product images in `pub/media/catalog/product` directory. You can temporary move the product directory to somewhere else (outside pub directory). then run again `bin/magento setup:upgrade`. This should be much faster this time. then move back the product directory to original location. This problem only occur when you run setup:upgrade command. This way may be not you expected but it's working for me.

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `Issue: Confirmed` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
@engcom-backlog-nazar Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-97435` were created

**Comment by marcusirgens** [comment]
I can confirm we are experiencing similar issues with several installations - a long chain of slow queries and locks caused by this step.

**Comment by oefterdal** [comment]
> @Flowlance I have an experience on this problem. I don't know exactly why this happened. Probably you have huge product images in `pub/media/catalog/product` directory. You can temporary move the product directory to somewhere else (outside pub directory). then run again `bin/magento setup:upgrade`. This should be much faster this time. then move back the product directory to original location. This problem only occur when you run setup:upgrade command. This way may be not you expected but it's working for me.

Hi I work with @Flowlance we don't have product images on this installation, all images are served via a CDN so no images are managed by Magento.

**Comment by Nazar65** [comment]
HI here, actually this is configuration issue about mysql.
Translation: by default, MySQL 5.5 will “meta-block” for 1 year!  In my humble opinion, this is a bug, especially given the various subtle and sometimes quiet ways that metadata locking can lock the server.  The default for innodb_lock_wait_timeout, by comparison, is 50 seconds.  That’s reasonable, but 31536000 is not.  I would only set a timeout, or any kind of wait or interval value, to such a high value to play a practical joke on someone.

```
session1 > start transaction;
Query OK, 0 rows affected (0.00 sec)
session1 > select * from test order by id;
+----+------+
| id | x    |
+----+------+
|  1 | foo  |
|  2 | bar  |
+----+------+
2 rows in set (0.00 sec)
session2 > ALTER TABLE test add column c char(32) default 'dummy_text';
session3 > show processlist;
+----+----------+-----------+------+---------+------+---------------------------------+-------------------------------------------------------------+
| Id | User     | Host      | db   | Command | Time | State                           | Info                                                        |
+----+----------+-----------+------+---------+------+---------------------------------+-------------------------------------------------------------+
|  1 | msandbox | localhost | test | Sleep   |  140 |                                 | NULL                                                        |
|  2 | msandbox | localhost | test | Query   |    3 | Waiting for table metadata lock | ALTER TABLE test add column c char(32) default 'dummy_text' |
|  3 | msandbox | localhost | test | Query   |    0 | NULL                            | show processlist                                            |
+----+----------+-----------+------+---------+------+---------------------------------+-------------------------------------------------------------+
3 rows in set (0.00 sec)

`session1 > start transaction;
Query OK, 0 rows affected (0.00 sec)
session1 > select * from test_2 order by id;
+----+------+------------+
| id | x    | c          |
+----+------+------------+
|  1 | foo  | dummy_text |
|  2 | bar  | dummy_text |
+----+------+------------+
2 rows in set (0.00 sec)
session6 > ALTER TABLE test_2 DROP COLUMN c;
session7 > select * from test_2 order by id;
session8 > select * from test_2 order by id;
session9 > select * from test_2 order by id;
session10 > select * from test_2 order by id;
session3 > show processlist;
+----+----------+-----------+------+---------+------+---------------------------------+----------------------------------+
| Id | User     | Host      | db   | Command | Time | State                           | Info                             |
+----+----------+-----------+------+---------+------+---------------------------------+----------------------------------+
|  1 | msandbox | localhost | test | Sleep   |  403 |                                 | NULL                             |
|  3 | msandbox | localhost | test | Query   |    0 | NULL                            | show processlist                 |
|  6 | msandbox | localhost | test | Query   |  229 | Waiting for table metadata lock | ALTER TABLE test_2 DROP COLUMN c |
|  7 | msandbox | localhost | test | Query   |  195 | Waiting for table metadata lock | select * from test_2 order by id |
|  8 | msandbox | localhost | test | Query   |  180 | Waiting for table metadata lock | select * from test_2 order by id |
|  9 | msandbox | localhost | test | Query   |  169 | Waiting for table metadata lock | select * from test_2 order by id |
| 10 | msandbox | localhost | test | Query   |   55 | Waiting for table metadata lock | select * from test_2 order by id |
+----+----------+-----------+------+---------+------+---------------------------------+----------------------------------+
7 rows in set (0.00 sec)`

**Comment by nicholasscottfish** [comment]
Hi @Nazar65, what was the fix for this, I seem to be experiencing the same issue on a site running 2.2.5; we're not currently in a position to upgrade.

**Comment by niteshv0909** [comment]
Why is is this closed without a fix or PR ? I am experiencing the same issue in Magento 2.3.

**Comment by francescorm** [comment]
i have got the same issue. Table Metadata Lock for trg_cataloginventory_stock_item_after_inser.

How do you have solved?

**Comment by convenient** [comment]
I just saw this on 2.3.4-p2

**Comment by marius-balteanu** [comment]
Same issue on 2.4.2.

**Comment by marius-balteanu** [comment]
@Nazar65, can you detail a little bit your last response, please? I'm not sure if your recommendation is to tweak the `innodb_lock_wait_timeout` value or no.

**Comment by sarron93** [comment]
@marius-balteanu , Hi I think that this solution will be support to you.

```php

namespace YourVendor\IndexerExtended\Setup;

use Magento\Framework\Indexer\IndexerInterfaceFactory;
use Magento\Framework\Setup\InstallDataInterface;
use Magento\Framework\Setup\ModuleContextInterface;
use Magento\Framework\Setup\ModuleDataSetupInterface;
use Magento\Framework\Indexer\ConfigInterface;
use Magento\Framework\Mview\ViewInterface;

class OverrideRecurringData extends \Magento\Indexer\Setup\RecurringData
{
    /**
     * @var IndexerInterfaceFactory
     */
    private $indexerFactory;

    /**
     * @var ConfigInterface
     */
    private $configInterface;

    /**
     * RecurringData constructor.
     *
     * @param IndexerInterfaceFactory $indexerFactory
     * @param ConfigInterface $configInterface
     */
    public function __construct(
        IndexerInterfaceFactory $indexerFactory,
        ConfigInterface $configInterface
    ) {
        $this->indexerFactory = $indexerFactory;
        $this->configInterface = $configInterface;
    }

    /**
     * {@inheritdoc}
     */
    public function install(ModuleDataSetupInterface $setup, ModuleContextInterface $context)
    {

        foreach (array_keys($this->configInterface->getIndexers()) as $indexerId) {
            $indexer = $this->indexerFactory->create()->load($indexerId);

            $subscriptionHash = $this->getSubscriptionHash($indexer->getView());

            if ($subscriptionHash == $indexer->getView()->getState()->getSubscribesHash()) {
                continue;
            }

            $indexer->getView()->getState()->setSubscribesHash($subscriptionHash);

            if ($indexer->isScheduled()) {
                $indexer->getView()->unsubscribe()->subscribe();
            }
        }
    }


    protected function getSubscriptionHash(ViewInterface $view)
    {
        $hashName = array_map(function ($subscription) {
            return $subscription['name'];
        }, $view->getSubscriptions());

        return md5(implode(',', $hashName));
    }
}
```
And di.xml
    <preference for="Magento\Indexer\Setup\RecurringData" type="YourVendor\IndexerExtended\Setup\OverrideRecurringData"/>

**Comment by Adel-Magebinary** [comment]
This still happens in 2.4.2

**Comment by otezz** [comment]
Hi @Nazar65 

I'm on magento 2.3.4 and mysql 5.7 aurora/rds
`SELECT @@innodb_lock_wait_timeout;` shows default value which is 50 and I get this issue

Any workaround recommendation for this?

**Comment by Nuranto** [comment]
This happens in 2.4.6 too. Only when servers have load.

**Comment by Nuranto** [comment]
Hi @engcom-Hotel,

Can you consider reopen this issue ?

The issue was confirmed, and then closed by @Nazar65 without any solution (we don't know which parameter he was refering to in his comment). 
Also, it would be better to fix this in code instead of custom settings in mysql/mariadb (which also seems risky). 

And as you can see, this is still affecting many of us.

**Comment by engcom-Bravo** [comment]
Hi @Nuranto,

We are reopening this issue.


Thanks.

**Comment by ingmarnet** [comment]
Hi, Any update with this? Still happening at 2.4.6-p2 Version

**Comment by pocallaghan-function** [comment]
As far as I can tell from debugging, the line that causes this problem, i.e. https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Indexer/Setup/Recurring.php#L124 appears to superfluous. The code that was merged from RecurringData into Recurring in [this commit](https://github.com/magento/magento2/commit/3ba26c7fd1ce1761714f125c2f508749c4f6126f) appears to be made completely redundant by the inclusion of `$installer->removeUnusedTriggers();`  in the https://github.com/magento/magento2/blob/2.4-develop/setup/src/Magento/Setup/Console/Command/UpgradeCommand.php#L148. The method `removeUnusedTriggers` does more than the name suggests and actually adds / updates triggers to, it does so however in a far more intelligent manner by determining if the triggers have actually changed, then updating only that trigger not all triggers for the view.

If anyone is aware of a reason the above statement is wrong and that the unsubscribe / subscribe has some other purpose, please let me know.

**Comment by engcom-Hotel** [comment]
Hello @pocallaghan-function,

First of all thanks for the detailed information.

@pocallaghan-function @Flowlance can you please provide us the latest steps to reproduce this issue. Please try to reproduce the issue in 2.4-develop and let us know if you are still able to get this issue.

Thanks

**Comment by warleyelias** [comment]
> As far as I can tell from debugging, the line that causes this problem, i.e. https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Indexer/Setup/Recurring.php#L124 appears to superfluous. The code that was merged from RecurringData into Recurring in [this commit](https://github.com/magento/magento2/commit/3ba26c7fd1ce1761714f125c2f508749c4f6126f) appears to be made completely redundant by the inclusion of `$installer->removeUnusedTriggers();` in the https://github.com/magento/magento2/blob/2.4-develop/setup/src/Magento/Setup/Console/Command/UpgradeCommand.php#L148. The method `removeUnusedTriggers` does more than the name suggests and actually adds / updates triggers to, it does so however in a far more intelligent manner by determining if the triggers have actually changed, then updating only that trigger not all triggers for the view.
> 
> If anyone is aware of a reason the above statement is wrong and that the unsubscribe / subscribe has some other purpose, please let me know.

@pocallaghan-function did you manage to find any solution on this case?

**Comment by pocallaghan-function** [comment]
@warleyelias I deployed this as a composer patch: 
```
diff --git a/Setup/Recurring.php b/Setup/Recurring.php
index 44d2687..6ee2ac5 100644
--- a/Setup/Recurring.php
+++ b/Setup/Recurring.php
@@ -120,11 +120,6 @@ class Recurring implements InstallSchemaInterface
                 $state->setStatus(StateInterface::STATUS_INVALID);
                 $state->save();
             }
-
-            $indexer = $this->indexerFactory->create()->load($indexerId);
-            if ($indexer->isScheduled()) {
-                $indexer->getView()->unsubscribe()->subscribe();
-            }
         }
     }
 }
```
We have not noticed any adverse affect.

**Comment by warleyelias** [comment]
@pocallaghan-function thank you, and what about the "removeUnusedTriggers" did you comment or change this to?

**Comment by engcom-Hotel** [comment]
Hello @warleyelias @pocallaghan-function,

Please try to reproduce this issue with the latest development branch i.e. 2.4-develop. And let us know of you are still able to reproduce the issue.

Thanks

**Comment by warleyelias** [comment]
@engcom-Hotel is not easy to replicate this problema using 2-4-develop once this problema start to occur when the application was deployed on production enviroment, that said I'm currently using the version 2.4.5-p5 and this issue was open since version 2.2.6 and the magento time even now could reproduce or resolve this issue?

**Comment by engcom-Hotel** [comment]
Closing this issue as duplicate of https://github.com/magento/magento2/issues/33386.

https://github.com/magento/magento2/issues/33386 is already confirmed.
